src/debug/profiler.rs

```diff
@@ -0,0 +1,183 @@
+use {
+    crate::state::AppState,
+    aws_config::meta::region::RegionProviderChain,
+    aws_sdk_s3::{config::Region, primitives::ByteStream, Client},
+    axum::{extract::State, response::IntoResponse, Json},
+    hyper::{HeaderMap, StatusCode},
+    once_cell::sync::Lazy,
+    serde::Deserialize,
+    std::{path::PathBuf, sync::Arc, time::Duration},
+    tokio::sync::Mutex,
+};
+
+static UPLOAD_CONTEXT: Lazy<Mutex<Option<UploadContext>>> = Lazy::new(|| Mutex::new(None));
+static PROFILER_LOCK: Lazy<Mutex<()>> = Lazy::new(|| Mutex::new(()));
+
+#[derive(Clone)]
+struct UploadContext {
+    bucket: String,
+    client: Client,
+}
+
+#[derive(Deserialize)]
+pub struct ProfilerPayload {
+    profile_duration: u64,
+}
+
+pub async fn init_upload_context(config: &crate::env::Config) {
+    let mut lock = UPLOAD_CONTEXT.lock().await;
+    *lock = create_upload_context(config).await;
+
+    tracing::warn!("profiler: initialized: {}", lock.is_some());
+}
+
+async fn create_upload_context(config: &crate::env::Config) -> Option<UploadContext> {
+    if let Some(bucket) = config.analytics.export_bucket.as_deref() {
+        tracing::warn!(%bucket, "profiler: initializing profiler...");
+
+        let region_provider = RegionProviderChain::first_try(Region::new("eu-central-1"));
+        let aws_config = aws_config::from_env().region(region_provider).load().await;
+
+        let client = Client::new(&aws_config);
+
+        Some(UploadContext {
+            client,
+            bucket: bucket.to_owned(),
+        })
+    } else {
+        None
+    }
+}
+
+pub async fn handler(
+    State(state): State<Arc<AppState>>,
+    headers: HeaderMap,
+    Json(payload): Json<ProfilerPayload>,
+) -> impl IntoResponse {
+    let auth = headers.get("Authorization");
+    if auth.is_none() {
+        return (hyper::StatusCode::UNAUTHORIZED, "Unauthorized");
+    }
+
+    if let Ok(auth) = auth.unwrap().to_str() {
+        if auth != state.config.debug.secret {
+            return (hyper::StatusCode::UNAUTHORIZED, "Unauthorized");
+        }
+    }
```

Comment:
security: If `.to_str()` returns `Err(..)` ([this happens if header value contains non-ASCII characters](https://docs.rs/hyper/latest/hyper/header/struct.HeaderValue.html)) then the secret isn't compared.

Furthermore, HeaderValue can be compared to strings, enabling this all to be simplified to:

```suggestion
    if auth != Some(state.config.debug.secret) {
        return (hyper::StatusCode::UNAUTHORIZED, "Unauthorized");
    }
```

Reply:
Are you sure about comparing HeaderValue to String? Doesn't seem to work for me

Reply:
According to these examples it can be done: https://docs.rs/hyper/latest/hyper/header/struct.HeaderMap.html#method.get

The [docs here](https://docs.rs/hyper/latest/hyper/header/struct.HeaderValue.html) say it is easy to compare to a string.